course_UvA-DL/initialization-and-optimization/Initialization_and_Optimization.py

```diff
@@ -0,0 +1,1152 @@
+#
+# If we assume now that $x$ is the output of a ReLU activation (from a previous layer, $x=max(0,\tilde{y})$), we can calculate the expectation as follows:
+#
+#
+# $$
+# \begin{split}
+#     \mathbb{E}[x^2] & =\mathbb{E}[\max(0,\tilde{y})^2]\\
+#                     & =\frac{1}{2}\mathbb{E}[{\tilde{y}}^2]\hspace{2cm}\tilde{y}\text{ is zero-centered and symmetric}\\
+#                     & =\frac{1}{2}\text{Var}(\tilde{y})
+# \end{split}$$
+#
+# Thus, we see that we have an additional factor of 1/2 in the equation, so that our desired weight variance becomes $2/d_x$.
+# This gives us the Kaiming initialization (see [He, K. et al.
+# (2015)](https://arxiv.org/pdf/1502.01852.pdf)).
+# Note that the Kaiming initialization does not use the harmonic mean between input and output size.
+# In their paper (Section 2.2, Backward Propagation, last paragraph), they argue that using $d_x$ or $d_y$ both lead to stable gradients throughout the network, and only depend on the overall input and output size of the network.
+# Hence, we can use here only the input $d_x$:
+
+
+# %%
+def kaiming_init(model):
+    for name, param in model.named_parameters():
+        if name.endswith(".bias"):
+            param.data.fill_(0)
+        elif name.startswith("layers.0"):  # The first layer does not have ReLU applied on its input
+            param.data.normal_(0, 1 / math.sqrt(param.shape[1]))
+        else:
+            param.data.normal_(0, math.sqrt(2) / math.sqrt(param.shape[1]))
+
+
+model = BaseNetwork(act_fn=nn.ReLU()).to(device)
+kaiming_init(model)
+visualize_gradients(model, print_variance=True)
+visualize_activations(model, print_variance=True)
+
+# %% [markdown]
+# The variance stays stable across layers.
+# We can conclude that the Kaiming initialization indeed works well for ReLU-based networks.
+# Note that for Leaky-ReLU etc., we have to slightly adjust the factor of $2$ in the variance as half of the values are not set to zero anymore.
+# PyTorch provides a function to calculate this factor for many activation
+# function, see `torch.nn.init.calculate_gain`
+# ([link](https://pytorch.org/docs/stable/nn.init.html#torch.nn.init.calculate_gain)).
+
+# %% [markdown]
+# ## Optimization
+#
+# Besides initialization, selecting a suitable optimization algorithm can be an important choice for deep neural networks.
+# Before taking a closer look at them, we should define code for training the models.
+# Most of the following code is copied from the previous tutorial, and only slightly altered to fit our needs.
+
+
+# %%
+def _get_config_file(model_path, model_name):
+    return os.path.join(model_path, model_name + ".config")
+
+
+def _get_model_file(model_path, model_name):
+    return os.path.join(model_path, model_name + ".tar")
+
+
+def _get_result_file(model_path, model_name):
+    return os.path.join(model_path, model_name + "_results.json")
+
+
+def load_model(model_path, model_name, net=None):
```

Comment:
seeing it in other notebooks, what do you think about having some common module for sharing?
**Pro**: do not repeat the same functions
**Con**: notebook wont be standalone any more

Reply:
I was also thinking about it, but we only have two notebooks that currently share some part of their code. I would prefer to keep the notebooks stand-alone and independent of each other, but I am also ok with making a shared module.